### PR TITLE
log-backup: use the span tree(instead of the naïve algorithm) to calculate the checkpoint 

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #N/A

Problem Summary:
We were using a naive algorithm to advance the checkpoint before:
it simply maintains a mapping from `timestamp` to `ranges`. Which would probably store too many radiant items.

### What is changed and how it works?

This PR added a new data structure named `spans.ValueSorted` in the code.
It contains a non-overlapped, full range tree as the "primary key" and a binary search tree with `timestamp` as key `range` as value for indexing. 
The primary tree has some characteristics:
- FULL: The union of all ranges it contains is the full range (the universal set).
- NON-OVERLAPPED: Any pair of ranges from it doesn't overlap.

For keeping these characteristics, the interface it provides is limited, when we inserting some ranges into it, it would "merge" those new ranges into its current state. 

An example here:

Each flush would generate a new checkpoint: usually, it updates the checkpoint of some range by the regions it flushed. In a more detailed view, you can imagine the key space as a line:
```
|___________________________________________________________________________|
""                                                                        inf
Then the checkpoint c of each range can be presented as:
|___________________________________________________________________________|
^-----------------^-----------------^-----------------^---------------------^
|      c = 42     |      c = 43     |     c = 45      |      c = 41         |
```
Once querying a checkpoint of some range, we simply choose the minimal of all its intersected ranges.
```
|___________________________________________________________________________|
^-----------------^-----------------^-----------------^---------------------^
|      c = 42     |      c = 43     |     c = 45      |      c = 41         |
    ^--------------------------^
    |   c = min(42, 43) = 42   |
```
A flush would update the checkpoint of a subrange. At this time, we fill the points in the range with max(new_checkpoint, old_checkpoint), we are going to call this step Merge, just like:
```
|___________________________________________________________________________|
^-----------------^-----------------^-----------------^---------------------^
|      c = 42     |      c = 43     |     c = 45      |      c = 41         |
                       ^--------------------------^
                       |      merge(c = 44)       |
```
Would Give:
```
|___________________________________________________________________________|
^-----------------^----^------------^-------------^---^---------------------^
|      c = 42     | 43 |   c = 44   |     c = 45      |      c = 41         |
                                    |-------------|
                                    Unchanged, because 44 < 45.
```

### Check List

Tests 

- [x] Unit test
- [] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/streamhelper/spans/value_sorted.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	v.MergeAll([]Valued{newItem})
+}
+
+func (v *ValueSortedFull) MergeAll(newItems []Valued) {
+	var overlapped []Valued
+	var inserted []Valued
+
+	for _, item := range newItems {
+		overlapped = overlapped[0:]
+		inserted = inserted[0:]
```
Comment (Leavrth) [inline br/pkg/streamhelper/spans/value_sorted.go:42]: may be
```
overlapped = overlapped[:0]
inserted = inserted[:0]
```

File: br/pkg/streamhelper/advancer.go [hunk 2]
```diff
@@ -214,121 +154,45 @@ func (c *CheckpointAdvancer) tryAdvance(ctx context.Context, rst RangesSharesTS)
...
-		})
-	}
-	err := eg.Wait()
+	err := c.tryAdvance(ctx, len(targets), func(i int) kv.KeyRange { return targets[i].Key })
 	if err != nil {
 		return 0, err
 	}
-	ts := c.cache.CheckpointTS()
+	ts := c.checkpoints.MinValue()
+	spans.Debug(c.checkpoints)
```
Comment (Leavrth) [inline br/pkg/streamhelper/advancer.go:192]: will it be removed before pr is merged? :eyes:

**Review by Leavrth — APPROVED** [review]

File: br/pkg/streamhelper/advancer.go (hunk 2 above)
Comment (YuJuncen) [reply br/pkg/streamhelper/advancer.go:192]: Well, that has been removed at the commit when subscription are supported, anyway I think we need to change it in this PR too...

**Review by joccau — COMMENTED** [review]
reset LGTM

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**